### ISSUE TITLE
[Python] `Table.to_pandas` fails to convert index `dtype` with a custom type mapper

### ISSUE THREAD

**Issue by j-bennet** [apache/arrow#34283]
### Describe the bug, including details regarding any error messages, version, and platform.

When providing a custom type mapper to `pyarrow.Table.to_pandas`, column dtypes are converted, but not index dtypes.

Example:

```python
import pyarrow as pa
import pandas as pd

df = pd.DataFrame({"s1": ["a", "c", "b"], "i1": [1, 2, 3], "s2": ["d", "e", "f"]}).set_index('s2')
tbl = pa.Table.from_pandas(df)
mapper = {pa.string(): pd.StringDtype("pyarrow")}.get
df2 = tbl.to_pandas(types_mapper=mapper)
print(f"{df2.dtypes}")
print(f"{df2.index.dtype = }")
```

This prints:

```
s1    string[pyarrow]
i1              int64
dtype: object
df2.index.dtype = dtype('O')
```

Column `s1` was mapped to `string[pyarrow]` dtype, however, the index dtype remained `object`. I'd expect the type mapper being used to convert the index `dtype` as well.

Platform: macOS
Pyarrow version: 11.0.0
Build: py310h89f3c6b_2_cpu
Channel: conda-forge

### Component(s)

Python

**Comment by AlenkaF** [comment]
Yes, I _think_ that the dtype of an index is never converted according to the `types_mapper` keyword when converting `pa.Table` or an array for that matter.

A single array gets converted to pandas series in `_array_like_to_pandas` with `pd.Series` which doesn't take into account the dtype of the index.

https://github.com/apache/arrow/blob/958f63502679c7074ff60c0d06dbb1a0be0a5cfb/python/pyarrow/array.pxi#L1649

This would be a good add on. What would be needed is to use pandas api to reset the series index with `.astype()` method after the code line linked above in [arrow/python/pyarrow/array.pxi](https://github.com/apache/arrow/blob/958f63502679c7074ff60c0d06dbb1a0be0a5cfb/python/pyarrow/array.pxi#L1649).

As for the Table, it gets converted with the use of pandas BlockManager and I am not sure how the desired dtype could be passed to the BlockManager axes in that case:

https://github.com/apache/arrow/blob/7828165f185b5ea2a3e76c06f9d4ba44263fd6dc/python/pyarrow/table.pxi#L4001-L4008

https://github.com/apache/arrow/blob/45918a90a6ca1cf3fd67c256a7d6a240249e555a/python/pyarrow/pandas_compat.py#L797-L823

**Comment by jorisvandenbossche** [comment]
Issue resolved by pull request 34445
https://github.com/apache/arrow/pull/34445
